### ISSUE TITLE
Get rotate boxes conner 

### ISSUE THREAD

**Issue by maudzung** [ming71/rotate-yolov3#12]
Hi @ming71 
I'm also working with rotate object detection, I'd like to introduce you to an efficient way to get coordinates of a rotated box, please refer to [get_corners](https://github.com/maudzung/Complex-YOLOv4-Pytorch/blob/ec420e007622175ee93e8a6043d6f2197a23dd67/src/data_process/kitti_bev_utils.py#L99).

Related to GIoU of 2 rotate boxes, did you find an efficient solution to calculate it? Can you please introduce it to me?

**Comment by ming71** [comment]
Thanks， I'll have a try on it.
IoU between rotate boxes can be obtained via various methods, such as [DOTA_devkit](https://github.com/CAPTAIN-WHU/DOTA_devkit/tree/master/poly_nms_gpu).But there is no clear definition of the  `GIoU` of 2 rotate boxes, I've tried to use similar idea to realize it via shapely, you can refer to this [code](https://github.com/ming71/toolbox/blob/master/rotation/SkewIou.py), hope it helps.

**Comment by maudzung** [comment]
Hi @ming71 
Thanks for your suggestion. Did you test the code for GIoU? From my point of view, the conners' points of a polygon (`union_poly` in your code) should be sorted/arranged in clockwise order (or vice versa) before going into `MultiPoint()`.
I'll try to apply the idea and share the output with you.

**Comment by ming71** [comment]
Hello, I  encountered the question you mentioned before,  and u r right, points must be sorted before calc giou.
I've solved it and put code [here ](https://github.com/ming71/toolbox/blob/master/rotation/order_points.py), have  a good luck!

**Comment by maudzung** [comment]
Thank you so much! I'll check it out.

**Comment by maudzung** [comment]
Hi @ming71 
I have found an effective way to calculate the GIoU for 2 rotated boxes. But the calculations are still conducted on CPU, not GPU.
I'd like to share it with you ([here](https://github.com/maudzung/Complex-YOLOv4-Pytorch/blob/7d7df357e7ee88d1f41f9add4c8bb32e7db5c3b1/src/utils/iou_rotated_boxes_utils.py#L101)). Although I tested it, can you have a check and also test it for me? Thank you.

**Comment by ming71** [comment]
Thanks, nice work!
But I'm busy with something else recently,  and I'll check the code for you when I'm free.

**Comment by maudzung** [comment]
Finally, **the GIoU loss can be done on GPU**. That's good because we don't need to `detach()` output tensors from the computational graph (this leads to backpropagation problems), and we also don't need to transfer data from GPU to CPU.
If you have time, please check it for me :-) I'm doing lonely, so I need other people to check and give me feedback. 
Thank you!
My implementation is [here](https://github.com/maudzung/Complex-YOLOv4-Pytorch/blob/b613d67e1187f000e3f7d15c49d1b213655b86b9/src/utils/iou_rotated_boxes_utils.py#L118)

**Comment by ming71** [comment]
> Finally, **the GIoU loss can be done on GPU**. That's good because we don't need to `detach()` output tensors from the computational graph (this leads to backpropagation problems), and we also don't need to transfer data from GPU to CPU.
> If you have time, please check it for me :-) I'm doing lonely, so I need other people to check and give me feedback.
> Thank you!
> My implementation is [here](https://github.com/maudzung/Complex-YOLOv4-Pytorch/blob/b613d67e1187f000e3f7d15c49d1b213655b86b9/src/utils/iou_rotated_boxes_utils.py#L118)

Hello, I feel so sorry for ignoring your comment here, your implementation is good! and I think it would be better if you replace overlaps calculation with CUDA codes, `shapely` is convenient but really slow, related codes can be found [here](https://github.com/ming71/toolbox/tree/master/rotation/overlaps/overlaps_cuda).
Besides, you can further speed up the calculation of overlaps by taking the minimum horizontal rectangle surrounding the rbox. This method is also used in [AerialDetection](https://github.com/dingjiansw101/AerialDetection), but the accuracy will drop slightly. Good luck to you!

**Comment by maudzung** [comment]
Thank you. Recently, I have developed a new repo that bases on the CenterNet's ideas. I thought the ideas can be applied to the rotated boxes detection tasks. There are several advantages:
- Don't need to find anchors
- Don't need to calculate the IoU of rotated boxes during training and inference. Hence the model can be faster in both training and inference phases. 

I'd like to introduce the work to you. My implementation is [here](https://github.com/maudzung/Super-Fast-Accurate-3D-Object-Detection). Please give me comments and suggestions to improve the repo. Many thanks once again.

**Comment by ming71** [comment]
Nice work! Have you checked it on datasets annotated with OBB? A center-net implementation for rotation detection have been reported on ECCV2020([PIoU loss](https://arxiv.org/abs/2007.09584)) but seems not good enough.
